pkg/test/harness.go

```diff
@@ -0,0 +1,280 @@
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// Harness loads and runs tests based on the configuration provided.
+type Harness struct {
+	T                 *testing.T
+	CRDDir            string
+	ManifestsDir      string
+	TestDirs          []string
+	StartControlPlane bool
+	StartKUDO         bool
+
+	managerStopCh chan struct{}
+	config        *rest.Config
+	env           *envtest.Environment
+}
+
+// LoadTests loads all of the tests in a given directory.
+func (h *Harness) LoadTests(dir string) ([]*Test, error) {
+	dir, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	tests := []*Test{}
+
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+
+		tests = append(tests, &Test{
+			Stages: []*Stage{},
+			Name:   file.Name(),
+			Dir:    filepath.Join(dir, file.Name()),
+		})
+	}
+
+	return tests, nil
+}
+
+// RunTestEnv starts a Kubernetes API server and etcd server for use in the
+// tests and returns the Kubernetes configuration.
+func (h *Harness) RunTestEnv() (*rest.Config, error) {
+	h.env = &envtest.Environment{}
+
+	config, err := h.env.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// Config returns the current Kubernetes configuration - either from the environment
+// or from the created temporary control plane.
+func (h *Harness) Config() (*rest.Config, error) {
+	if h.config != nil {
+		return h.config, nil
+	}
+
+	var err error
+
+	if h.StartControlPlane {
+		h.config, err = h.RunTestEnv()
+	} else {
+		h.config, err = config.GetConfig()
+	}
+
+	return h.config, err
+}
+
+// RunKUDO starts the KUDO controllers and installs the required CRDs.
+func (h *Harness) RunKUDO(crdPath string) error {
+	config, err := h.Config()
+	if err != nil {
+		return err
+	}
+
+	if h.CRDDir != "" {
+		crds, err := envtest.InstallCRDs(config, envtest.CRDInstallOptions{
+			Paths:              []string{h.CRDDir},
+			ErrorIfPathMissing: true,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = envtest.WaitForCRDs(config, crds, envtest.CRDInstallOptions{
+			Paths:              []string{h.CRDDir},
+			ErrorIfPathMissing: true,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	mgr, err := manager.New(config, manager.Options{})
+	if err != nil {
+		return err
+	}
+
+	if err = apis.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
+	}
+
+	if err = controller.AddToManager(mgr); err != nil {
+		return err
+	}
+
+	if err = webhook.AddToManager(mgr); err != nil {
+		return err
+	}
+
+	h.managerStopCh = make(chan struct{})
+	go mgr.Start(h.managerStopCh)
+
+	return nil
+}
+
+// InstallManifests recurses over ManifestsDir to install all resources defined in YAML manifests.
+func (h *Harness) InstallManifests() error {
```

Comment:
[minor] - I am not sure if 'harness' is the right package term for methods like InstallManifests (I'd expect it under utils).
But a refactor / renaming will be under the purview of kudo maintainers.